proposals/2023-06-13_created-timestamp.md

```diff
@@ -0,0 +1,152 @@
+
+#### Prometheus Protobuf
+
+Since Prometheus already supports scraping endpoints with Prometheus protobuf and knowing that adding the created timestamp field is trivial, it makes Prometheus protobuf the prefered option for initial implementation:
+
+```proto
+message Counter {
+  optional double   value    = 1;
+  optional Exemplar exemplar = 2;
+  optional google.protobuf.Timestamp    created_timestamp_ms = 3;
+}
+
+message Summary {
+  optional uint64   sample_count = 1;
+  optional double   sample_sum   = 2;
+  repeated Quantile quantile     = 3;
+  optional google.protobuf.Timestamp    created_timestamp_ms = 4;
+}
+
+message Histogram {
+  // …
+  repeated double positive_count    = 14; // Absolute count of each bucket.
+  optional google.protobuf.Timestamp  created_timestamp_ms = 15;
+}
+```
+
+Once the protocol is extended, Prometheus parser can easily identify created timestamps and ingest as mentioned in the [How](#how) section.
+
+#### OpenMetrics Protobuf
+
+Follows the same semantics as mentioned in [Prometheus Protobuf](#prometheus-protobuf), but Prometheus still isn't able to parse it.
+
+The task here is to simply implement the missing parser.
+
+#### OpenMetrics text
+
+For OpenMetrics text the situation is more complex. For OM text format, the _created timestamps series would be normally ingested as a normal series. However, for this solution we propose to capture the timestamp from the series and treat it as the synthetic zero. Then we propose to remove the artificial series to avoid further metric collisions, semantic differences (series vs created timestamp metadata) and consistency with other exposition formats.
+
+The removal of a special metric is a **breaking change**. However, we still propose doing it to stop the spread of this confusing pattern in the community (with adequate warning). The current usage of created metrics is low and users are [actually not adopting OM](https://github.com/prometheus/client_python/tree/v0.17.1#disabling-_created-metrics) for that reason. Thus allowing users to opt-out (users can set if they actually want to preserve those metrics) seems like a good way forward.
+
+Note that, the conversion of  _created series to created timestamp append will not be a trivial, or efficient code. We have to check various metrics, and buffer appends of potentially related metrics. Ideally, this will go away from the next version of OpenMetrics text (outside of the scope of this proposal).
+
+
+## Alternatives
+
+### Inject sentinel values
+
+Instead of injecting zero samples, a sentinel value, e.g. _CreatedNaN_ could be ingested intead to avoid confusion when noticing that not all samples follow the configured scrape interval.
+
+It would also help remote-write ingestors to easily identify created timestamps in a batch of samples.
+
+It would require refactoring of rate-like functions, where sentinel values would need to be replaced with zeros.
```

Comment:
I think zero injection might mess with "count_over_time" and other horizontal aggregations maybe, so probably this also needs to be addressed in the query layer right?

Reply:
Yes, it will change the outcome of `count_over_time`, that's why it has to be behind a feature flag.

The question is more if it's a change for the better or for the worse.

Reply:
It's a good point though, it should be mentioned in the proposal

Reply:
added this problem to the "Feature flag" section